add example of a good docstring for defaults and examples (#12356)
scikit-learn/scikit-learn: doc/developers/contributing.rst

    SelectKBest : Select features based on the k highest scores.
    SelectFpr : Select features based on a false positive rate test.

* When documenting the parameters and attributes, have the following in mind:

>     1. Do not use optional. Use `default=`. `str {'a', 'b'} or float, default=1.0`

Reviewer: (Not sure?)
```suggestion
    1. Do not use optional. Use `default=`.
```

    2. Python basic types. (`bool` instead of `boolean`)
    3. When defining 1-D shape, use parenthesis. `array-like shape=(n_samples,), None, default=None`.

>     4. When defining 2-D shape, use parenthesis. `array-like shape=(n_samples, n_features)`.

Reviewer: merge with 3?

>     4. When defining 2-D shape, use parenthesis. `array-like shape=(n_samples, n_features)`.

Reviewer 2: What about the case that the array can be both 1-D and 2-D (e.g. output `y_pred`)

    5. str with multiple options: `input: {'log', 'squared', 'multinomial'}`
    6. Only use `or` for separating types. `float, int or None, default=None`

>     7. Only use comma to separate types. Information such as `shape` or `string`

Reviewer: I'm not sure what you mean by that: you also use a comma in `{'a', 'b'}`?

Also this is rst, not markdown ;)

    options are defined together. Here is an extreme example:

    ```

>     {'a', 'b'}, float, int, array-like shape=(n_samples,) or None, default=None

Reviewer 2: If `shape=` is a standard then I am fine with it. But I would rather find the following easier to parse visually.

```python
"""
array-like of shape (n_samples,)
{array-like, sparse matrix or dataframe} of shape (n_samples, n_features)
"""
```

Reviewer 3: I am okay with `of shape`. I would go with:

```
{array-like, sparse matrix, dataframe} of shape (n_samples, n_features)
```

(There is no "or" before "dataframe")

    ```

>     8. When supporting array-like and sparse matrix use:

Reviewer: Why having a special case for this?

`array-like or sparse matrix shape=(n_samples, n_features)`

would work too?

Reviewer 2: Is there a reason to not have any separator between the `matrix type` and `shape`.

Reviewer 2: Since that we can have dataframe as well, we could make this point more generic.

    SelectKBest : Select features based on the k highest scores.
    SelectFpr : Select features based on a false positive rate test.

* When documenting the parameters and attributes, here is a list of some good
  examples::

    some_param : {'hello', 'goodbye'}, bool or int, default=True
	    The parameter description goes here, which can be either a string
        literal (either `hello` or `goodbye`), a bool, or an int. The default
        value is True.

    array_parameter : {array-like, sparse matrix, dataframe} of shape (n_samples, n_features) or (n_samples,)

>         This parameter accepts a data in either of the mentioned forms, with one

Reviewer 4: ```suggestion
        This parameter accepts data in either of the mentioned forms, with one
```

    SelectKBest : Select features based on the k highest scores.
    SelectFpr : Select features based on a false positive rate test.

* When documenting the parameters and attributes, here is a list of some
  well-formatted examples::

    n_clusters : int, default=3
        The number of clusters detected by the algorithm.

    some_param : {'hello', 'goodbye'}, bool or int, default=True
        The parameter description goes here, which can be either a string
        literal (either `hello` or `goodbye`), a bool, or an int. The default
        value is True.

>     array_parameter : {array-like, sparse matrix, dataframe} of shape (n_samples, n_features) or (n_samples,)

Reviewer 4: uhh did we have a discussion of "of shape" vs "shape=". Did someone count that?

Reviewer 4: "of shape": 513 lines; "shape=" 338 lines; "shape =" 1316 lines. "of shape =": 162 lines, "of shape=" 1 line.

These are regex matches, so non-exclusive, so "of shape" includes "of shape =".
So we have 350 "of shape (...)" and 1115 "shape = (..)"

Reviewer 3: Yes `shape=` would result in a smaller diff. Since this PR will result in many docstring changes, we should decide based on which is subjectively better.

(I have been +0.25 for `shape=` because it saves a few more characters.) Do you have a opinion on this?

Reviewer 4: My preference is shape= but I don't mind that much.

Reviewer 4: we are often short on space in that line.

``array-like of shape (n_samples,)`` vs ``array-like, shape=(n_samples,)`` saves 2 characters. Not sure if that's a good argument?

Reviewer 3: I was thinking of pushing it to:

```py
array-like shape=(n_samples,)
```

Extreme savings of 3 characters.

@glemaitre I know you prefer `of shape` because it is easier to parse. Would you be okay with `shape=`?

Reviewer 4: I don't like not having the comma :P

Reviewer 4: ![image](https://user-images.githubusercontent.com/449558/62652933-77221c80-b92a-11e9-8105-49e2f2183cbd.png)

Reviewer: I don't think saving 2-3 chars is a major improvement here. Yes, that avoids having to use `\` in a few cases, but the rendered html are what are important, and they don't care.

Reviewer 5: > My preference is shape= but I don't mind that much.

I also prefer shape= a little bit, I merge this because I think it's not important as long as the whole repo is consistent.

Author: we can fix those cases that don't comply with this with a `sed`, so I don't think the number of instances of each case are important that much anyway. I kinda liked not having the comma cause the comma is separating other things:

`int, array-like of shape (n_samples,), or None, default=None` (note that we don't have the oxford comma in our guide now, but I rather have it)

Reviewer 2: > I don't like not having the comma :P

I prefer `of shape` if there is no comma. I find it more difficult to visually parse it.
I also find it closer to literal English --- even if we are writing documentation for expert computer scientists (I should not troll about Matlab post).

Anyway, I am going with the consensus (since it is merged it seems we are going for that one?)

Reviewer 4: Ok, fine with keeping it.
@adrinjalali feel free to go ahead and do the sed and then fix all the merge conflicts :P
If whis is the way we want to go, we should probably do the sed.

Reviewer 4: So are we doing this?

Author: I guess so, I'll submit a PR.

    SelectKBest : Select features based on the k highest scores.
    SelectFpr : Select features based on a false positive rate test.

> * When documenting the parameters and attributes, here is a list of some good

Reviewer: ```suggestion
* When documenting the parameters and attributes, here is a list of some well-formatted
```

  examples::

    n_clusters : int, default=3
        The number of clusters detected by the algorithm.

    some_param : {'hello', 'goodbye'}, bool or int, default=True
        The parameter description goes here, which can be either a string
        literal (either `hello` or `goodbye`), a bool, or an int. The default
        value is True.

    array_parameter : {array-like, sparse matrix, dataframe} of shape (n_samples, n_features) or (n_samples,)
        This parameter accepts data in either of the mentioned forms, with one
        of the mentioned shapes. The default value is
        `np.ones(shape=(n_samples,))`.

In general have the following in mind:

>     1. Python basic types. (`bool` instead of `boolean`)

Reviewer: ```suggestion
    1. Use Python basic types. (`bool` instead of `boolean`)
```

    2. Use parenthesis for defining shapes: `array-like of shape (n_samples,)`

>     or `array-like of shape (n_samples, n_features)`

Reviewer: This list doesn't render properly, you'll need to indent the new lines (please checked rendered doc)

>     3. str with multiple options: `input: {'log', 'squared', 'multinomial'}`

Reviewer: ```suggestion
    3. For strings with multiple options, use brackets: `input: {'log', 'squared', 'multinomial'}`
```

    4. 1D or 2D data can be a subset of
    `{array-like, ndarray, sparse matrix, dataframe}`. Note that `array-like`
    can also be a `list`, while `ndarray` is explicitly only a `numpy.ndarray`.
    An example supporting two of them would be::

>         `{array-like, sparse matrix} shape=(n_samples, n_features)`

Reviewer: ```suggestion
        {array-like, sparse matrix} shape=(n_samples, n_features)
```

Though honestly I'd even remove the example since we have one above.

    SelectKBest : Select features based on the k highest scores.
    SelectFpr : Select features based on a false positive rate test.

* When documenting the parameters and attributes, here is a list of some good
  examples::

    some_param : {'hello', 'goodbye'}, bool or int, default=True
	    The parameter description goes here, which can be either a string
        literal (either `hello` or `goodbye`), a bool, or an int. The default
        value is True.

    array_parameter : {array-like, sparse matrix, dataframe} of shape (n_samples, n_features) or (n_samples,)
        This parameter accepts a data in either of the mentioned forms, with one
        of the mentioned shapes.

In general have the following in mind:

>     1. Do not use optional. Use `default=`.

Reviewer: Might as well just remove to not give users bad ideas?

    2. Python basic types. (`bool` instead of `boolean`)

>     3. When defining 1-D shape, use parenthesis. `array-like of shape (n_samples,), None, default=None`.

Reviewer: I'd merge 3 and 4 with just 

> Use parenthesis for defining shapes: `array-like of shape (n_samples,)` or `array-like of shape (n_samples, n_features)`

    4. When defining 2-D shape, use parenthesis. `array-like of shape (n_samples, n_features)`.
    5. str with multiple options: `input: {'log', 'squared', 'multinomial'}`

>     6. Only use `or` for separating types. `float, int or None, default=None`

Reviewer: I'd remove 6 and 7 since we don't even follow these in the examples above.

We use or for separating shapes, and we use commas everywhere in the examples, not just for separating types.

Reviewer 3: I agree with remove 6 and 7.

    7. Only use comma to separate types. Information such as `shape` or `string`
    options are defined together. Here is an extreme example::

        {'a', 'b'}, float, int, array-like shape=(n_samples,) or None, default=None

>     8. Array or matrix-like data can be a subset of

Reviewer: I'd say 1D or 2D data can be a subset of... 

smarse -> sparse

Might be worth explaining that array-like allows lists while ndarray is only for numpy arrays.
